62_CAD/DrawResourcesFiller.cpp

```diff
@@ -309,256 +360,683 @@ void DrawResourcesFiller::drawFontGlyph(
+			cachedImageRecord->arrayIndex = InvalidTextureIndex;
 		}
 	}
-	return success;
+
+	
+	// cached or just inserted, we update the lastUsedFrameIndex
+	cachedImageRecord->lastUsedFrameIndex = currentFrameIndex;
+
+	assert(cachedImageRecord->arrayIndex != InvalidTextureIndex); // shouldn't happen, because we're using LRU cache, so worst case eviction will happen + multi-deallocate and next next multi_allocate should definitely succeed
+	return (cachedImageRecord->arrayIndex != InvalidTextureIndex);
 }
 
-bool DrawResourcesFiller::finalizeLineStyleCopiesToGPU(SIntendedSubmitInfo& intendedNextSubmit)
+bool DrawResourcesFiller::queueGeoreferencedImageCopy_Internal(image_id imageID, const StreamedImageCopy& imageCopy)
 {
-	bool success = true;
-	// Copy LineStyles
-	uint32_t remainingLineStyles = currentLineStylesCount - inMemLineStylesCount;
-	SBufferRange<IGPUBuffer> stylesRange = { sizeof(LineStyle) * inMemLineStylesCount, sizeof(LineStyle) * remainingLineStyles, gpuDrawBuffers.lineStylesBuffer };
-	if (stylesRange.size > 0u)
-	{
-		const LineStyle* srcLineStylesData = reinterpret_cast<LineStyle*>(cpuDrawBuffers.lineStylesBuffer->getPointer()) + inMemLineStylesCount;
-		if (m_utilities->updateBufferRangeViaStagingBuffer(intendedNextSubmit, stylesRange, srcLineStylesData))
-			inMemLineStylesCount = currentLineStylesCount;
-		else
-		{
-			// TODO: Log
-			success = false;
-		}
+	auto& vec = streamedImageCopies[imageID];
+	vec.emplace_back(imageCopy);
+	return true;
+}
+
+// TODO[Przemek]: similar to other drawXXX and drawXXX_internal functions that create mainobjects, drawObjects and push additional info in geometry buffer, input to function would be a GridDTMInfo
+// We don't have an allocator or memory management for texture updates yet, see how `_test_addImageObject` is being temporarily used (Descriptor updates and pipeline barriers) to upload an image into gpu and update a descriptor slot (it will become more sophisticated but doesn't block you)
+void DrawResourcesFiller::drawGridDTM(
+	const float64_t2& topLeft,
+	float64_t height,
+	float64_t width,
+	float gridCellWidth,
+	const DTMSettingsInfo& dtmSettingsInfo,
+	SIntendedSubmitInfo& intendedNextSubmit)
+{
+	GridDTMInfo gridDTMInfo;
+	gridDTMInfo.topLeft = topLeft;
+	gridDTMInfo.height = height;
+	gridDTMInfo.width = width;
+	gridDTMInfo.gridCellWidth = gridCellWidth;
+
+	if (dtmSettingsInfo.mode & E_DTM_MODE::OUTLINE)
+	{
+		const bool isOutlineStippled = dtmSettingsInfo.outlineStyleInfo.stipplePatternSize > 0;
+		gridDTMInfo.outlineStipplePatternLengthReciprocal = isOutlineStippled ? dtmSettingsInfo.outlineStyleInfo.reciprocalStipplePatternLen : 0.0f;
 	}
-	return success;
+
+	setActiveDTMSettings(dtmSettingsInfo);
+	beginMainObject(MainObjectType::GRID_DTM);
+
+	uint32_t mainObjectIdx = acquireActiveMainObjectIndex_SubmitIfNeeded(intendedNextSubmit);
+	assert(mainObjectIdx != InvalidMainObjectIdx);
+
+	addGridDTM_Internal(gridDTMInfo, mainObjectIdx);
```

Comment:
you need to follow this pattern to handle possible overflow submit (in case there is not enough mem for GridDTMInfo)
```c++
	if (!addImageObject_Internal(info, mainObjIdx))
	{
		// single image object couldn't fit into memory to push to gpu, so we submit rendering current objects and reset geometry buffer and draw objects
		submitCurrentDrawObjectsAndReset(intendedNextSubmit, mainObjIdx);
		bool success = addImageObject_Internal(info, mainObjIdx);
		assert(success); // this should always be true, otherwise it's either bug in code or not enough memory allocated to hold a single image object 
	}
```